src/Data/TaskMap.hs

```diff
@@ -88,14 +96,14 @@ memo
     -> (k -> IO v)
         -- ^ an action that is used to produce the value if the key isn't in the map.
     -> IO v
-memo tm@(TaskMap var) k task = do
-    -- TODO: should we insert another lookup here? It depends on how optimistic
-    -- we are that the lookup is successful.
-
-    a <- modifyMVarMasked var $ \m -> case HM.lookup k m of
-        Nothing -> do
-            !a <- asyncWithUnmask $ \umask -> umask (task k) `finally` delete tm k
-            m' <- evaluate $ HM.insert k a m
-            return (m', a)
-        (Just !a) -> return (m, a)
-    wait a
+memo tm@(TaskMap var) k task = bracket query cancel wait
+  where
+    query = do
+        -- TODO: should we insert another lookup here? It depends on how optimistic
+        -- we are that the lookup is successful.
+        modifyMVarMasked var $ \m -> case HM.lookup k m of
+            Nothing -> do
+                !a <- asyncWithUnmask $ \umask -> umask (task k) `finally` delete tm k
+                !m' <- evaluate $ HM.insert k a m
```

Comment:
Is it meaningful to use a bang pattern with the result of an `evaluate` call?

Reply:
It is probably not meaningful, though not 100% sure. After having stumbled several times over strictness annotation semantics -- and also received contradictory advice from people that I consider experts in the field -- I kinda gave up trying to be smart on this. Also for other maintainers, I think, it's better to be explicit than assume that every Haskeller can quickly identify redundant strictness annotations. Instead, I tent to just put the bang when I won't things to be strict, not matter if it's redundant or not.

Reply:
For instance, is this line equivalent here to `let !m' = HM.insert k a m`? It seems like it would be the same in this case. But I am completely sure.

Reply:
For hashmap, keys, values and key arguments are evaluated to WHNF before entry into the map, so the `!` here is to force an already spine-strict structure.

Reply:
> Is it meaningful to use a bang pattern with the result of an `evaluate` call?

No, it isn't. 

> For instance, is this line equivalent here to `let !m' = HM.insert k a m`? It seems like it would be the same in this case. But I am completely sure.

In this case yes, but not always. `evaluate` sequences evaluation with respect to execution of IO actions, whereas `!` via `seq` manufactures data dependencies. The difference is that IO actions are executed strictly sequentially and data dependencies form a partial order. For example, a "sequence" of strict bindings one after another in a `do`-block are not ordered with respect to one another unless there are data dependencies between them, so the compiler is free to reorder them with respect to one another. This can't happen with `evaluate`.

Reply:
I agree that it is redundant in this case. My argument is that code is about readability. In this case the bang pattern just documents that this value should be in whnf. This is the case independent of whether it is already or if it because of the bang pattern.

Adding redundant bang patterns does't make programs less readable (in my opinion). They actually can make the program easier to read, because the reader sees the intent of the code author without having to study the intrinsics of Haskells evaluation rules.

Our code should be accessible and maintainable not only by the most seasoned Haskell programmers. While we should expect a certain level of expertise from devs, I don't think that we can or should expect instantaneous judgment of redundancy of bang patterns in all cases.

Reply:
I propose that discussions about bang patterns in PRs focused on the question whether strictness is desirable or not for a value and not on the question whether they are redundant. I think, the former is what is important the latter is a style question.